anywidget/_descriptor.py

```diff
@@ -0,0 +1,615 @@
+
+    return _default_set_state
+
+
+# ------------- Psygnal support --------------
+
+
+def _get_psygnal_signal_group(obj: object) -> psygnal.SignalGroup | None:
+    """Look for a psygnal.SignalGroup on the obj."""
+    psygnal = sys.modules.get("psygnal")
+    if psygnal is None:
+        return None
+
+    # most likely case: signal group is called "events"
+    events = getattr(obj, "events", None)
+    if isinstance(events, psygnal.SignalGroup):
+        return events
+
+    # try exhaustive search
+    with contextlib.suppress(
+        (AttributeError, RecursionError, TypeError)
+    ):  # pragma: no cover
+        for attr in vars(obj).values():
+            if isinstance(attr, psygnal.SignalGroup):
+                return attr
+
+
+def _connect_psygnal(obj: object, send_state: Callable) -> Callable | None:
+    """Check if an object has a psygnal.SignalGroup, and connect it to send_state.
+
+    Returns
+    -------
+    disconnect : Callable | None
+        A callable that disconnects the psygnal.SignalGroup from send_state, or None
+        if no psygnal.SignalGroup was found.
+    """
+    events = _get_psygnal_signal_group(obj)
+
+    if events is not None:
+
+        @events.connect
+        def _on_psygnal_event(event: psygnal.EmissionInfo):
+            send_state({event.signal.name})
+
+        def _disconnect():
+            events.disconnect(_on_psygnal_event)
+
+        return _disconnect
+    return None
+
+
+# ------------- Traitlets support --------------
+
+
+def _is_traitlets_object(obj: Any) -> TypeGuard[traitlets.HasTraits]:
+    """Return `True` if an object is an instance of traitlets.HasTraits."""
+    traitlets = sys.modules.get("traitlets")
+    return isinstance(obj, traitlets.HasTraits) if traitlets is not None else False
+
+
+# a tag that can be added to a traitlet to indicate that it should be synced
+# this apparently comes from ipywidgets, not traitlets
+_TRAITLETS_SYNC_FLAG = "sync"
+
+
+# TODO: decide about usage of "sync" being opt-in or opt-out
+# users of traitlets who *don't* use ipywidgets might be surprised when their
+# state isn't being synced without opting in.
+
+
+def _get_traitlets_state(obj: traitlets.HasTraits) -> dict:
+    """Get the state of a traitlets.HasTraits instance."""
+    return obj.trait_values(**{_TRAITLETS_SYNC_FLAG: True})
+
+
+def _connect_traitlets(obj: object, send_state: Callable) -> Callable | None:
+    """Check if an object has a psygnal.SignalGroup, and connect it to send_state.
+
+    Returns
+    -------
+    disconnect : Callable | None
+        A callable that disconnects the psygnal.SignalGroup from send_state, or None
+        if no psygnal.SignalGroup was found.
+    """
+    if not _is_traitlets_object(obj):
+        return None
+
+    obj_ref = weakref.ref(obj)
+
+    @obj.observe
+    def _on_trait_change(change: dict):
+        name = change["name"]
+        if obj_ref().trait_metadata(name, _TRAITLETS_SYNC_FLAG):
+            send_state({name})
```

Comment:
I guess an alternative here would be to only `observe` to the known traits with `sync=True` metadata (rather than have this condition every change event.

```python
obj.observe(
    lambda change: send_state({ change["name"] }),
    names=list(obj.traits(sync=True).keys()),
)
```

This would of course prevent users from _dynamically_ setting traits. 

```python
counter = CounterWidget()
counter.add_traits(name=traitlets.Unicode().tag(sync=True))
```

So maybe the above is preferred.

Reply:
ah good point.  yeah I'm torn.

Reply:
> with a slight preference for only observing a set of names to begin with

Same. tbh I can't really think of a case where dynamically setting state to sync with the front end makes sense. The JS is defined statically, I think it's fair to also make the traits be defined statically.

Reply:
all set